tests/nn/data_parallel/test_fsdp_overlap.py

```diff
@@ -0,0 +1,249 @@
+        has_params = all_gather_cycles > 0
+        model = _create_model(fsdp_config, compute_cycles, has_params)
+
+        # if gpu_id == 0:
+        #    print(model)
+
+        # Get the input and sets the input's requires_grad to True because
+        # we have a fake compute in the forward pass.
+        batch = torch.rand(1).cuda()
+        batch.requires_grad = True
+
+        # Save the original torch.distributed.all_gather function since we will
+        # patch it to include an artificial delay.
+        orig_all_gather = torch.distributed.all_gather
+
+        # We run 20 iterations but only collect timing data from the minimal 10
+        # data points because nondeterministic system events can disturb the timing.
+        cpu_iter = Min10()
+        cpu_wait = Min10()
+        gpu_compute = Min10()
+        gpu_total = Min10()
+        for _ in range(20):
+            # Two events for measuring the overall time.
+            e1 = Event(enable_timing=True)
+            e2 = Event(enable_timing=True)
+
+            cpu_start = time.process_time()
+
+            all_gather_called = False
+
+            def _delayed_all_gather(*args, **kwargs):
+                nonlocal all_gather_called
+                all_gather_called = True
+                torch.cuda._sleep(all_gather_cycles)
+                return orig_all_gather(*args, **kwargs)
+
+            # forward
+            # Even though both e1 & e2 are on the compute stream, since
+            # compute depends on all_gather, e2-e1 includes all_gather time.
+            e1.record()
+            with patch("torch.distributed.all_gather", _delayed_all_gather):
+                out = model(batch)
+                if has_params and world_size > 1:
+                    assert all_gather_called
+                else:
+                    assert not all_gather_called
+            e2.record()
+
+            # backward
+            out.backward()
+            if torch_version() >= (1, 7, 0):
+                model.zero_grad(set_to_none=True)
+            else:
+                for p in model.parameters():
+                    p.grad = None
+
+            cpu_iter_time = time.process_time() - cpu_start
+
+            # wait for gpu
+            out.item()
+            cpu_wait_for_gpu_time = time.process_time() - cpu_start - cpu_iter_time
+
+            # get sum of compute time
+            times = []
+            for mod in model.modules():
+                if not isinstance(mod, Layer):
+                    continue
+                times.append(mod.get_time())
+
+            # get gpu compute + all_gather time
+            overall_gpu_time = e1.elapsed_time(e2)
+
+            # print(f"rank {rank}", cpu_iter_time, cpu_wait_for_gpu_time, sum(times), "vs", e1.elapsed_time(e2))
+            cpu_iter.add(cpu_iter_time)
+            cpu_wait.add(cpu_wait_for_gpu_time)
+            gpu_compute.add(sum(times))
+            gpu_total.add(overall_gpu_time)
```

Comment:
Copying the discussion from here: https://github.com/facebookresearch/fairscale/pull/672#discussion_r629034005

> I seem to see all of them 1-2 ms: 
> <img width="1215" alt="image" src="https://user-images.githubusercontent.com/24926999/117699722-03a1e480-b17a-11eb-9e6b-ac86fb5279bf.png">

Interesting! I see several hundred ms on my local V100 machine. It's very strange that CI is reporting such different results... I wonder if something weird is happening in CI?

Reply:
My local test shows that it is passing with very good values. clearly the overlapping is there. :-)